packages/@react-aria/numberfield/src/useNumberField.ts

```diff
@@ -96,81 +124,187 @@ export function useNumberField(props: NumberFieldProps, state: NumberFieldState,
+   */
+  let numberFormatter = useNumberFormatter(formatOptions);
+  let intlOptions = useMemo(() => numberFormatter.resolvedOptions(), [numberFormatter]);
+  let hasDecimals = intlOptions.maximumFractionDigits > 0;
+  let inputMode: 'decimal' | 'numeric' | 'text' = hasDecimals ? 'decimal' : 'numeric';
+  if (state.minValue < 0) { // iOS - neither allows negative signs, so use full keyboard
+    inputMode = 'text';
+  }
+
+  let {onChange, onKeyDown, onPaste} = useSelectionControlledTextfield({setValue: state.setValue, value: state.inputValue, isFocused}, ref);
 
-  let domProps = filterDOMProps(props, {labelable: true});
   let {labelProps, inputProps} = useTextField(
-    mergeProps({
+    {
+      label,
       autoFocus,
       isDisabled,
       isReadOnly,
       isRequired,
       validationState,
-      value: isFocused ? inputValue : textValue,
+      value: state.inputValue,
       autoComplete: 'off',
       'aria-label': props['aria-label'] || null,
       'aria-labelledby': props['aria-labelledby'] || null,
       id: inputId,
       placeholder: formatMessage('Enter a number'),
-      type: 'text',
-      onChange: state.setValue
-    }), ref);
+      type: 'text', // Can't use type="number" because then we can't have things like $ in the field.
+      inputMode,
+      onChange,
+      onKeyDown,
+      onPaste
+    }, ref);
 
-  const inputFieldProps = mergeProps(focusProps, spinButtonProps, inputProps);
+  const inputFieldProps = mergeProps(
+    spinButtonProps,
+    inputProps,
+    focusProps,
+    {
+      onWheel,
+      // override the spinbutton role, we can't focus a spin button with VO
+      role: null,
+      'aria-roledescription': formatMessage('Spin button number field'),
+      'aria-valuemax': null,
+      'aria-valuemin': null,
+      'aria-valuenow': null,
+      'aria-valuetext': null
+    }
+  );
   return {
-    numberFieldProps: mergeProps(domProps, {
+    numberFieldProps: {
       role: 'group',
       'aria-disabled': isDisabled,
-      'aria-invalid': validationState === 'invalid'
-    }),
+      'aria-invalid': validationState === 'invalid' ? 'true' : undefined,
+      ...focusWithinProps
+    },
     labelProps,
     inputFieldProps,
     incrementButtonProps,
     decrementButtonProps
   };
 }
+
+
+let useSelectionControlledTextfield = ({setValue, value, isFocused}: {setValue: (string) => void, value: string, isFocused: RefObject<boolean>}, ref: RefObject<HTMLInputElement>) => {
+  /**
+   * Selection is to track where the cursor is in the input so we can restore that position + some offset after render.
+   * The reason we have to do this is because the user is not as limited when the field is empty, the set of allowed characters
+   * is at the maximum amount. Once a user enters a numeral, we determine the system and close off the allowed set.
+   * This means we can't block the values before they make it to state and cause a render, thereby moving selection to an
+   * undesirable location.
```

Comment:
it does briefly change because i don't prevent anyone from typing anything, but i do rapidly remove. That's enough to confuse the browsers selection position